### DISCUSSION TITLE
Usage instructions — discrete allocation and CombinatorialPurgedCV

### DISCUSSION THREAD

**Discussion by rriski** [skfolio/skfolio#15] [General]
Hey, thanks for developing this great library! I'm trying to utilize skfolio to build a personal investment portfolio. I'm using the provided example notebooks, specifically the HRP vs HERC [notebook](https://skfolio.org/auto_examples/5_clustering/plot_3_hrp_vs_herc.html#sphx-glr-auto-examples-5-clustering-plot-3-hrp-vs-herc-py). This example uses the [`CombinatorialPurgedCV`](https://skfolio.org/generated/skfolio.model_selection.CombinatorialPurgedCV.html#skfolio.model_selection.CombinatorialPurgedCV) class to run cross validation with multiple test paths from different training folds combinations. 

Now for my first question: is `CombinatorialPurgedCV` purely intended to validate model behavior, i.e. see the distributions of different measures (Sharpe, CVaR, etc) to determine the "stability" of the model? Or should I pick one path (the best one?) from the output of `CombinatorialPurgedCV` and use the asset allocation of that path?

Then another question / note: I was experimenting with the [PyPortfolioOpt](https://github.com/robertmartin8/PyPortfolioOpt) lib before finding skfolio. They have a nice `[DiscreteAllocation`](https://pyportfolioopt.readthedocs.io/en/latest/Postprocessing.html?highlight=discrete#pypfopt.discrete_allocation.DiscreteAllocation) helper to convert asset weights, latest price data and target portfolio size into discrete number of assets to buy. Are there any plans of incorporating something like that into skfolio?

My final question relates to to the [prices_to_returns](https://github.com/skfolio/skfolio/blob/main/src/skfolio/preprocessing/_returns.py#L10) helper. If I have pricing data such as this one, where stock B has entered the stock market after A

   | A | B
-- | -- | --
2014-01-07 | 19.36 | NaN
2014-01-08 | 19.29 | NaN
2014-01-09 | 19.50 | NaN
2014-01-10 | 19.52 | NaN
2014-01-11 | 19.50 | NaN
... | ...
2023-12-29 | 41.87 | 6.00
2024-01-02 | 42.13 | 6.02
2024-01-03 | 41.96 | 6.11
2024-01-04 | 42.50 | 6.09
2024-01-05 | 42.84 | 6.10

...then `prices_to_returns` will trim both A and B values to start from the first available date where B has data:

A | B
-- | --
2022-12-02 | -0.003124 | 0.011236
2022-12-05 | 0.010184 | -0.013699
2022-12-07 | -0.009306 | 0.003086
2022-12-08 | -0.011742 | -0.007692
2022-12-09 | 0.008317 | 0.000000
... | ...
2023-12-29 | 0.002634 | -0.008264

Is there a way around this? I would like to train a model with a portfolio that includes B, but having the data truncated like that makes it more difficult to build a good model, as so much information is lost.

**Comment by HugoDelatte** [comment]
Hello, thank you for your feedback and your PR! 

Indeed, `CombinatorialPurgedCV` is used for cross-validation and helps you analyze the distribution of risk measures for different combinations of train/test folds. Once you are satisfied with your model, you should fit it to the most recent data using the same time window as you did in your cross-validation (for example, 252 days).

Yes, we could add a utility function that optimally discretizes the assets units. The reason it was not included is that the package was initially used for institutional investment, for which unit rounding is negligible compared to the notional invested. But I recognise that for small investments it may be beneficial. 

Regarding the `prices_to_returns` function, it doesn't fill missing data because that's a very complex problem, especially over long periods. So that's a problem left to the investor. In your example, I would suggest to optimize only after the introduction of B.

**Reply by CarloNicolini** [reply]
It is possible to do a post-processing step on asset weights to help retail investors to allocate following constraints dictated by minimum amount indeed.

Something similar is described in PyportfolioOpt under the DiscreteAllocation class.
Otherwise designing a smart set of constraints could also be of help.

**Reply by jmlocatelli** [reply]
Hi. Thanks for this very usefull library.
Regarding the last question about prices_to_return calculation for stocks with no data during certain period I would like to ask.
Depending on the market, many stocks are listed and delisted during the long test period. So many stocks will have no data for some periods.
So, during a walk forward test for long periods, is it possible just to drop stocks with no data only for the train/test fold where all or part of data doesn't exist? Like pre-selection.
And, as market regime changes, can pre-selection be applied separetely for each train/test fold during walk forward?
Many thanks,

**Reply by HugoDelatte** [reply]
By using a `transformer` in a Pipeline that filters empty columns (assets that are not yet listed or delisted during that period) would work for the `fit` step of the cross-validation but fail at the `predict` step if the listed assets are not the same between the train and test sets.

I wonder how this problem could be solved, as it requires additional information from the investor. For example, let's assume that you run an Equal-Weighted model. On the train set, suppose that all three assets A, B, and C are listed; the computed weights are 1/3 for each. Now, you apply this allocation to the test set, but C has been delisted. If you simply assign 1/3 to A and 1/3 to B, you now have a different model (far from equal-weighted). In practice, when this happens, the investor finds a replacement for asset C and reruns the model on the new dataset.

So, instead of running cross-validation on the full sparse universe, you would run it on multiple complete subsets.

Let me know what you think.

**Reply by jmlocatelli** [reply]
I trade in a market that has a relatively small number of assets. So I want to use the maximum amount of available information to analyze the models. For this, automatic selection of assets is more convenient. The most complicated criterion is to remove assets that do not have all the data available during the train/test period. Since I always use walk forward optimization to evaluate any strategy, this filtering has to be performed for each train/test fold separately, to filter out assets without daily liquidity, newly listed assets, and delisted assets.
Thinking about possible solutions using skfolio, I came up with some ideas:
- create a transformer to pre-select the assets involved in each train/test fold, train/test the model for this fold, and aggregate the result with that of other folds.
- replace NaNs with a small value, create a group with the name of these assets for each fold, and run the model using linear constraints considering zero weights for this group.

I came up with this solution that seems to work. I still need to check it better. Suggestions and comments are welcome.

```
print("\nprices\n",prices)

# optimization model
model = MeanRisk(
    prior_estimator=EmpiricalPrior(
        covariance_estimator=EmpiricalCovariance(nearest=True)
        )
    )

# train/test grid
tscv = WalkForward(train_size=252, test_size=60)

# run thru WFO
population = MultiPeriodPortfolio()                      # population of portfolios

for train_idx, test_idx in tscv.split(prices):
    # split train/test fold
    train_data = prices.iloc[train_idx]
    test_data = prices.iloc[test_idx]

    # remove columns/assets with any missing data  
    nan_cols_train_data = train_data.columns[train_data.isnull().any()].tolist()
    nan_cols_test_data = test_data.columns[test_data.isnull().any()].tolist()
    train_data = train_data.drop(columns=nan_cols_train_data, errors='ignore')
    train_data = train_data.drop(columns=nan_cols_test_data, errors='ignore')
    test_data = test_data.drop(columns=nan_cols_train_data, errors='ignore')
    test_data = test_data.drop(columns=nan_cols_test_data, errors='ignore')
    print("\nprices_train_data\n",train_data)
    print("\nprices_test_data\n",test_data)

    # convert to returns
    train_data = prices_to_returns(train_data)
    test_data = prices_to_returns(test_data)
    print("\nreturns_train_data\n",train_data)
    print("\nreturns_test_data\n",test_data)

    # fit and predict
    model.fit(train_data)
    print("\nmodel.weights_\n",model.weights_)
    pred_model = model.predict(test_data)
    print("\npred_model:\n",pred_model.summary())
    population.append(pred_model)

# consolidated results
print("\npopulation_pred_model:\n",population.summary())
fig = population.plot_cumulative_returns()
show(fig)
```

**Reply by CarloNicolini** [reply]
Have you taken a look at [this](https://portfoliooptimizer.io/blog/managing-missing-asset-returns-in-portfolio-analysis-and-optimization-backfilling-through-residuals-recycling/) approach instead?

This blog post is based on the papers by [Stambaugh](https://www.sciencedirect.com/science/article/pii/S0304405X97000202) and [Page](https://www.tandfonline.com/doi/abs/10.2469/faj.v69.n1.3)

@HugoDelatte
Probably something that we could take a look at,  in terms of a sklearn imputer. Possibly inherit directly from `sklearn.imputer._BaseImputer`?

**Reply by jmlocatelli** [reply]
Hi, thank you for the valuable references. Very good approaches, I really didn't think about. They suggest to apply something like interpolation (my words) to fill the gaps. But in case of stock prices, long periods with missing data occurrs with listed and delisted stocks, and very sparse data for stocks with low liquidity. As my appoach is to filter stocks separatedly for each relatively small train/test fold during walk forward, I understand that the impact of discarding entire colunms only in that fold is minimized.
My other chalenge is to learn how to do Hyper-Parameters Tuning. I came from Fortran and punched cards era so Python and OOP are, I would say, still mistic for me.

**Reply by HugoDelatte** [reply]
@CarloNicolini I agree, skfolio would benefit from an `Imputer` based on Stambaugh's paper. We should implement it.

@jmlocatelli I see, as opposed to my initial example, you also use the test set (so the entire train/test fold) to filter the assets. It means that this logic needs to be implemented "manually" as you did or inside the [`cross_val_predict`](https://github.com/skfolio/skfolio/blob/4982e274c9bce4813e3ef1e2f10dd6a299876324/src/skfolio/model_selection/_validation.py#L25) function. A pre-selection `transformer` wouldn't work because it can only transform data from the input X which would be the train set and has no information about the test set. 

Concerning your last question about Hyper-Parameters Tuning, you can start with these resources:
* https://skfolio.org/user_guide/hyper_parameters_tuning.html
* https://skfolio.org/auto_examples/1_mean_risk/plot_8_regularization.html#hyper-parameter-tuning
* https://skfolio.org/auto_examples/6_ensemble/plot_1_stacking.html#parameter-tuning
